### PR TITLE
Fix incorrect length of strings containing unicode characters in window mode.

### PR DESCRIPTION
**Pull request by flhp**
There was a problem displaying workspace names/ window titles etc. if they contained unicode characters such as symbols, smileys, greek letters, etc.
Here's an example:
rofi_old
Here's how it should look (in my opinion):
rofi_new

G_Strings seem to incorrectly report their own length, however the g_utf8_strlen function returns the correct value. I've added a check for this situation and appended the correct number of missing whitespaces.

If you'd prefer a different solution, I'd be happy to do it differently.
Thanks!

### CODE REVIEW

File: source/dialogs/window.c [hunk 1]
```diff
@@ -631,14 +631,21 @@ struct arg
 
 static void helper_eval_add_str ( GString *str, const char *input, int l, int max_len )
 {
-    int nc = g_utf8_strlen ( input, -1 );
-    if ( l != 0 && nc > l ) {
-        nc = l;
-        int bl = g_utf8_offset_to_pointer ( input, nc ) - input;
+    GString* g_input = g_string_new( input );
```
Comment (DaveDavenport) [inline source/dialogs/window.c]: Why create a new GString for this?
Also where is it free'ed?

File: source/dialogs/window.c [hunk 2]
```diff
@@ -631,14 +631,21 @@ struct arg
...
+    int input_len = g_utf8_strlen( g_input->str, -1 );
+    if ( l != 0 && input_len > l ) {
+        input_len = l;
+        int bl = g_utf8_offset_to_pointer ( input, input_len ) - input;
         g_string_append_len ( str, input, bl );
     }
-    else{
+    else {
+        int d = str->len - g_utf8_strlen( str->str, -1 );
```
Comment (DaveDavenport) [inline source/dialogs/window.c]: This d is not used, and overwritten in line 644?

File: source/dialogs/window.c [hunk 3]
```diff
@@ -631,14 +631,21 @@ struct arg
...
         g_string_append_printf ( str, "%-*s", l ? l : max_len, input ? input : "" );
+        d = g_input->len - input_len;
+        l = l ? l : max_len;
+        if ( d > 0 ) {
+            g_string_append_printf( str, "%*s", input_len + d <= l ? d : l - input_len, "" );
```
Comment (DaveDavenport) [inline source/dialogs/window.c]: x= input_len+d
d = g_input->len - input_len
x = input_len + g_input->len  - input_len = g_input->len

File: source/dialogs/window.c [hunk 4]
```diff
@@ -479,7 +479,7 @@ static void _window_mode_load_data ( Mode *sw, unsigned int cd )
                 else {
                     c->wmdesktopstr = g_strdup ( "" );
                 }
-                pd->wmdn_len = MAX ( pd->wmdn_len, strlen ( c->wmdesktopstr ) );
+                pd->wmdn_len = MAX ( pd->wmdn_len, g_utf8_strlen ( c->wmdesktopstr, -1) );
```
Comment (DaveDavenport) [inline source/dialogs/window.c]: shouldn't we do this for all fields, not just the desktop string.

**Comment by DaveDavenport** [comment]
Does this one do what you want?
```patch
diff --git a/source/dialogs/window.c b/source/dialogs/window.c
index 1b353f4..83852ee 100644
--- a/source/dialogs/window.c
+++ b/source/dialogs/window.c
@@ -294,17 +294,17 @@ static client* window_client ( ModeModePrivateData *pd, xcb_window_t win )
     if ( c->title == NULL ) {
         c->title = window_get_text_prop ( c->window, XCB_ATOM_WM_NAME );
     }
-    pd->title_len = MAX ( c->title ? strlen ( c->title ) : 0, pd->title_len );
+    pd->title_len = MAX ( c->title ? g_utf8_strlen ( c->title, -1 ) : 0, pd->title_len );
 
     c->role      = window_get_text_prop ( c->window, netatoms[WM_WINDOW_ROLE] );
-    pd->role_len = MAX ( c->role ? strlen ( c->role ) : 0, pd->role_len );
+    pd->role_len = MAX ( c->role ? g_utf8_strlen ( c->role, -1 ) : 0, pd->role_len );
 
     cky = xcb_icccm_get_wm_class ( xcb->connection, c->window );
     xcb_icccm_get_wm_class_reply_t wcr;
     if ( xcb_icccm_get_wm_class_reply ( xcb->connection, cky, &wcr, NULL ) ) {
         c->class     = rofi_latin_to_utf8_strdup ( wcr.class_name, -1 );
         c->name      = rofi_latin_to_utf8_strdup ( wcr.instance_name, -1 );
-        pd->name_len = MAX ( c->name ? strlen ( c->name ) : 0, pd->name_len );
+        pd->name_len = MAX ( c->name ? g_utf8_strlen ( c->name, -1 ) : 0, pd->name_len );
         xcb_icccm_get_wm_class_reply_wipe ( &wcr );
     }
 
@@ -441,7 +441,7 @@ static void _window_mode_load_data ( Mode *sw, unsigned int cd )
                  && !client_has_window_type ( c, xcb->ewmh._NET_WM_WINDOW_TYPE_DESKTOP )
                  && !client_has_state ( c, xcb->ewmh._NET_WM_STATE_SKIP_PAGER )
                  && !client_has_state ( c, xcb->ewmh._NET_WM_STATE_SKIP_TASKBAR ) ) {
-                pd->clf_len = MAX ( pd->clf_len, ( c->class != NULL ) ? ( strlen ( c->class ) ) : 0 );
+                pd->clf_len = MAX ( pd->clf_len, ( c->class != NULL ) ? ( g_utf8_strlen ( c->class, -1 ) ) : 0 );
 
                 if ( client_has_state ( c, xcb->ewmh._NET_WM_STATE_DEMANDS_ATTENTION ) ) {
                     c->demands = TRUE;
@@ -479,7 +479,7 @@ static void _window_mode_load_data ( Mode *sw, unsigned int cd )
                 else {
                     c->wmdesktopstr = g_strdup ( "" );
                 }
-                pd->wmdn_len = MAX ( pd->wmdn_len, strlen ( c->wmdesktopstr ) );
+                pd->wmdn_len = MAX ( pd->wmdn_len, g_utf8_strlen ( c->wmdesktopstr, -1 ) );
                 if ( cd && c->wmdesktop != current_desktop ) {
                     continue;
                 }
@@ -631,14 +631,27 @@ struct arg
 
 static void helper_eval_add_str ( GString *str, const char *input, int l, int max_len )
 {
-    int nc = g_utf8_strlen ( input, -1 );
-    if ( l != 0 && nc > l ) {
-        nc = l;
-        int bl = g_utf8_offset_to_pointer ( input, nc ) - input;
-        g_string_append_len ( str, input, bl );
-    }
-    else{
-        g_string_append_printf ( str, "%-*s", l ? l : max_len, input ? input : "" );
+    // g_utf8 does not work with NULL string.
+    const char *input_nn = input ? input : "";
+    // Both l and max_len are in characters, not bytes.
+    int        nc = g_utf8_strlen ( input_nn, -1 );
+    int spaces = 0;
+    if ( l == 0 ) {
+        spaces = MAX ( 0, max_len - nc );
+        g_string_append ( str, input_nn );
+    }
+    else {
+        if ( nc > l ) {
+            int bl = g_utf8_offset_to_pointer ( input_nn, l ) - input_nn;
+            g_string_append_len ( str, input_nn, bl );
+        }
+        else {
+            spaces = l - nc;
+            g_string_append ( str, input_nn );
+        }
+    }
+    while ( spaces-- ) {
+        g_string_append_c ( str, ' ' );
     }
 }
 static gboolean helper_eval_cb ( const GMatchInfo *info, GString *str, gpointer data )
```

**Comment by flhp** [comment]
You're right, I was a bit sloppy.
That code does indeed fix the problem!

**Comment by DaveDavenport** [comment]
Thanks!

**Pull request closed without merging**